rnote-compose/src/shapes/arrow.rs

```diff
@@ -0,0 +1,194 @@
+use kurbo::PathEl;
+use na::Rotation2;
+use p2d::bounding_volume::Aabb;
+use serde::{Deserialize, Serialize};
+
+use crate::helpers::Vector2Helpers;
+use crate::shapes::ShapeBehaviour;
+use crate::transform::TransformBehaviour;
+
+use super::Line;
+
+/// All doc-comments of this file and [`ArrowBuilder`] rely on the following
+/// graphic:
+/// ```
+///         tip
+///         /|\
+///        / | \
+///       /  |  \
+///    lline |  rline
+///          |
+///          |
+///          |
+///         start
+/// ```
+/// Where `lline`, `tip`, `start` and `rline` represent a vector of the arrow.
+#[derive(Debug, Clone, Serialize, Deserialize, Default)]
+#[serde(default, rename = "arrow")]
+pub struct Arrow {
+    /// The start of the arrow
+    pub start: na::Vector2<f64>,
+
+    /// The tip of the arow
+    pub tip: na::Vector2<f64>,
+}
+
+impl TransformBehaviour for Arrow {
+    fn translate(&mut self, offset: na::Vector2<f64>) {
+        self.start += offset;
+        self.tip += offset;
+    }
+
+    fn rotate(&mut self, angle: f64, center: na::Point2<f64>) {
+        let isometry = {
+            let mut isometry = na::Isometry2::identity();
+            isometry.append_rotation_wrt_point_mut(&na::UnitComplex::new(angle), &center);
+            isometry
+        };
+
+        self.start = (isometry * na::Point2::from(self.start)).coords;
+        self.tip = (isometry * na::Point2::from(self.tip)).coords;
+    }
+
+    fn scale(&mut self, scale: na::Vector2<f64>) {
+        self.start = self.start.component_mul(&scale);
+        self.tip = self.tip.component_mul(&scale);
+    }
+}
+
+impl ShapeBehaviour for Arrow {
+    fn bounds(&self) -> Aabb {
+        self.internal_compute_bounds(None)
+    }
+
+    fn hitboxes(&self) -> Vec<Aabb> {
+        let n_splits = super::hitbox_elems_for_shape_len((self.tip - self.start).norm());
+
+        self.split(n_splits)
+            .into_iter()
+            .map(|line| line.bounds())
+            .collect()
+    }
+}
+
+impl Arrow {
+    /// The tip lines (rline & lline) default length, is the actual length when
+    /// no stroke width is associated with the arrow, if there is, it is used as
+    /// the base length for the actual width calculation.
+    const TIP_LINES_DEFAULT_LENGTH: f64 = 10.0;
+
+    /// The angle for `rline` and `lline` to the stem of the arrow.
+    const TIP_LINES_STEM_ACUTE_ANGLE: f64 = (13.0 / 16.0) * std::f64::consts::PI;
```

Comment:
sorry, this actually needs to be named `..OBTUSE_ANGLE` .

Reply:
xD

Done.

Reply:
and, you know, all occurrences have to be changed

Reply:
*damn it*

*bonk*

Reply:
*now* it should be done (sorry)